### PR TITLE
Port MYFACES-4433 to MyFaces Implementations (2.2, 2.3, and 3.0)

### LINKED ISSUE #20950: Memory Leak with JSF's ViewScopeContextualStorage (MYFACES-4433)
**Describe the bug**  
There is a small memory leak in the JSF features. CDI ViewScoped beans are stored within the ViewScopeContextualStorage object, and these beans are removed at the end of their lifecycle. Each new view creates another ViewScopeContextualStorage to store the beans. However, each ViewScopeContextualStorage is not deleted after the view changes. Instead, it's only deleted at session expiration. 

Note that this leak is very small, but may still had up over time until the session expires. 

**Steps to Reproduce**  
Applications uses CDI ViewScoped beans

**Expected behavior**  
No memory leak should occur.  ViewScopeContextualStorage should be removed entirely after the beans it holds are destroyed. 

**Diagnostic information:**  
 - OpenLiberty Version: Up and including 22.0.0.5
 - Affected feature(s) jsf-2.2, jsf-2.3, and faces-3.0
 - Java Version: N/A
 - server.xml configuration: N/A

**Additional context**  
Open Source Issue: https://issues.apache.org/jira/browse/MYFACES-4433

### PR DESCRIPTION
**Pull request by volosied**
- fixes #20950 for jsf-2.2, jsf-2.3, and faces-3.0

- Creates new test to verify session size after two views are created and destroyed

### CODE REVIEW

**Comment by volosied** [comment]
#build  #spawn.fullfat.buckets=com.ibm.ws.jsf.2.2_fat

File: dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22MiscellaneousTests.java [hunk 1]
```diff
@@ -516,4 +519,49 @@ public void testCustomValueBindingTag() throws Exception {
...
+
+            URL url = JSFUtils.createHttpUrl(jsf22MiscellaneousServer, "ViewScopeLeak", "index.xhtml");
+
+            String size1 = "1";
+            String size2 = "2";
+
+            // Runs twice. Goes to index.xhtml to create a view new. Goes to new page (invalidate) to end the view.
+            // invalidate.xhtml will list size of WELD_S#0 via SessionSizeHelper. 
+            // The size should be the same for runs.
+            // WELD_S#0 is the attribute where the ViewScope related storage is saved via CDI.
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22MiscellaneousTests.java:542]: 1) Should it read: `create a new view`?
2) The size should be the same for **both** runs?

File: dev/com.ibm.ws.jsf.2.2_fat/test-applications/ViewScopeLeak.war/src/com/ibm/ws/jsf22/fat/viewscopedleak/ViewScopedCDIBean.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+/**
+ * A bean that is used to increment and decrement a value.
+ * 
+ * This bean acts both as a listener as well as a regular bean.
+ * 
+ * 
+ * There is also a method to cause an error.
+ * 
+ * @author Paul Nicolucci
+ */
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsf.2.2_fat/test-applications/ViewScopeLeak.war/src/com/ibm/ws/jsf22/fat/viewscopedleak/ViewScopedCDIBean.java:28]: This looks like a copy paste class, are there changes to warrant a copyright date update etc? Should this be a new class?

Comment (volosied) [reply dev/com.ibm.ws.jsf.2.2_fat/test-applications/ViewScopeLeak.war/src/com/ibm/ws/jsf22/fat/viewscopedleak/ViewScopedCDIBean.java:28]: Oops. Yes, I copied this class over from an existing test application.  Let me fix it.

File: dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22MiscellaneousTests.java (hunk 1 above)
Comment (volosied) [reply dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22MiscellaneousTests.java:542]: I mean to say`new view`  I'll fix it.

And yes, the size should be the same for _both_ runs. The ViewScoped Beans are helper classes should be deleted once the views end.  The appropriate session attribute would then be the same.

Comment (volosied) [reply dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22MiscellaneousTests.java:542]: Updated to say both.

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.2_fat

**Review by pnicolucci — APPROVED** [review]

**Comment by pnicolucci** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.2_fat

**Comment by pnicolucci** [comment]
#libby

**Pull request merged**